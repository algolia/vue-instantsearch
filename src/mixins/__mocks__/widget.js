let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
