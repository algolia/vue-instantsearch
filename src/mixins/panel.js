import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
