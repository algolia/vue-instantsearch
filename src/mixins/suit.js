import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
