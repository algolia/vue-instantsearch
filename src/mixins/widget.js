const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
