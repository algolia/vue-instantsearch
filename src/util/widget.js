export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
