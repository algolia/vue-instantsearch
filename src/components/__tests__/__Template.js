import { mount } from '@vue/test-utils';
import Template from '../__Template.vue';
import { __setState } from '../../mixins/widget';
jest.mock('../../mixins/widget');

it('renders correctly', () => {
  __setState({
    hits: ['yo', 'how', 'are', 'you', 'doing', '?'],
  });
  const wrapper = mount(Template);
  expect(wrapper.html()).toMatchSnapshot();
});

// ☑️ add another rendering test if it's different given the propsData

it('behaves correctly', () => {
  __setState({
    refine: jest.fn(),
  });
  const wrapper = mount(Template);
  const button = wrapper.find('button');
  button.trigger('click');
  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith('hi');
});
