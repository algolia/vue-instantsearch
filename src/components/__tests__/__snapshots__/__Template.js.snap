// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div class="ais-Template">
  <button class="ais-Template-button">
    example refine
  </button>
  <span class="json-tree json-tree-root">
    <span class="json-tree-deep">
      <span class="json-tree-row json-tree-expando">
        <span class="json-tree-indent">
          &nbsp;
        </span>
        <span class="json-tree-sign">
          -
        </span>
        <span class="json-tree-indent">
          &nbsp;
        </span>
        <span class="json-tree-open">
          {
        </span>
        <span class="json-tree-collapsed"
              style="display: none;"
        >
          &nbsp;/*&nbsp;1 item&nbsp;*/&nbsp;
        </span>
        <span class="json-tree-close"
              style="display: none;"
        >
          }
        </span>
        <span class="json-tree-comma"
              style="display: none;"
        >
          ,
        </span>
        <span class="json-tree-indent">
          &nbsp;
        </span>
      </span>
      <span class="json-tree-deeper">
        <span class="json-tree">
          <span class="json-tree-deep">
            <span class="json-tree-row json-tree-expando">
              <span class="json-tree-indent">
                &nbsp;
              </span>
              <span class="json-tree-sign">
                -
              </span>
              <span class="json-tree-indent">
                &nbsp;
              </span>
              <span class="json-tree-indent">
                &nbsp;
              </span>
              <span class="json-tree-indent">
                &nbsp;
              </span>
              <span class="json-tree-key">
                "hits"
              </span>
              <span class="json-tree-colon">
                :&nbsp;
              </span>
              <span class="json-tree-open">
                [
              </span>
              <span class="json-tree-collapsed"
                    style="display: none;"
              >
                &nbsp;/*&nbsp;6 items&nbsp;*/&nbsp;
              </span>
              <span class="json-tree-close"
                    style="display: none;"
              >
                ]
              </span>
              <span class="json-tree-comma"
                    style="display: none;"
              >
                ,
              </span>
              <span class="json-tree-indent">
                &nbsp;
              </span>
            </span>
            <span class="json-tree-deeper">
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;yo&quot;"
                  >
                    "yo"
                  </span>
                  <span class="json-tree-comma">
                    ,
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;how&quot;"
                  >
                    "how"
                  </span>
                  <span class="json-tree-comma">
                    ,
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;are&quot;"
                  >
                    "are"
                  </span>
                  <span class="json-tree-comma">
                    ,
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;you&quot;"
                  >
                    "you"
                  </span>
                  <span class="json-tree-comma">
                    ,
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;doing&quot;"
                  >
                    "doing"
                  </span>
                  <span class="json-tree-comma">
                    ,
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
              <span class="json-tree">
                <span class="json-tree-row">
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                  <span class="json-tree-value json-tree-value-string"
                        title="&quot;?&quot;"
                  >
                    "?"
                  </span>
                  <span class="json-tree-indent">
                    &nbsp;
                  </span>
                </span>
              </span>
            </span>
            <span class="json-tree-row">
              <span class="json-tree-ending">
                <span class="json-tree-indent">
                  &nbsp;
                </span>
                <span class="json-tree-indent">
                  &nbsp;
                </span>
                <span class="json-tree-indent">
                  &nbsp;
                </span>
                <span class="json-tree-indent">
                  &nbsp;
                </span>
                <span class="json-tree-indent">
                  &nbsp;
                </span>
                <span class="json-tree-close">
                  ]
                </span>
                <span class="json-tree-indent">
                  &nbsp;
                </span>
              </span>
            </span>
          </span>
        </span>
      </span>
      <span class="json-tree-row">
        <span class="json-tree-ending">
          <span class="json-tree-indent">
            &nbsp;
          </span>
          <span class="json-tree-indent">
            &nbsp;
          </span>
          <span class="json-tree-indent">
            &nbsp;
          </span>
          <span class="json-tree-close">
            }
          </span>
          <span class="json-tree-indent">
            &nbsp;
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;
